apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.UnitTests/SessionPoolTests.cs

```diff
@@ -0,0 +1,443 @@
+                pool.ReleaseToPool(client.Object, sessions[1]);
+                pool.ReleaseToPool(client.Object, sessions[2]);
+
+                Assert.Equal(
+                    2, pool.GetPoolSize(
+                        client.Object, s_defaultName.ProjectId,
+                        s_defaultName.InstanceId, s_defaultName.DatabaseId));
+            }
+        }
+
+        [Fact]
+        public async Task MaxActiveViolationThrows()
+        {
+            using (var pool = new SessionPool())
+            {
+                pool.Options.WaitOnResourcesExhausted = false;
+                pool.Options.MaximumActiveSessions = 2;
+                bool exceptionThrown = false;
+
+                var client = CreateMockClient();
+                await pool.CreateSessionFromPoolAsync(
+                        client.Object, s_defaultName.ProjectId,
+                        s_defaultName.InstanceId, s_defaultName.DatabaseId, null, CancellationToken.None)
+                    .ConfigureAwait(false);
+                await pool.CreateSessionFromPoolAsync(
+                        client.Object, s_defaultName.ProjectId,
+                        s_defaultName.InstanceId, s_defaultName.DatabaseId, null, CancellationToken.None)
+                    .ConfigureAwait(false);
+                try
+                {
+                    await pool.CreateSessionFromPoolAsync(
+                        client.Object, s_defaultName.ProjectId,
+                        s_defaultName.InstanceId, s_defaultName.DatabaseId, null, CancellationToken.None)
+                        .ConfigureAwait(false);
+                }
+                catch (RpcException e)
+                {
+                    Assert.Equal(StatusCode.ResourceExhausted, e.Status.StatusCode);
+                    exceptionThrown = true;
+                }
+
+                Assert.True(exceptionThrown);
+            }
+        }
+
+        [Fact]
+        public async Task MaxActiveViolationBlocks()
+        {
+            using (var pool = new SessionPool())
+            {
+                pool.Options.WaitOnResourcesExhausted = true;
+                pool.Options.MaximumActiveSessions = 2;
+
+                var client = CreateMockClient();
+                var session1 = await pool.CreateSessionFromPoolAsync(
+                    client.Object, s_defaultName.ProjectId,
+                    s_defaultName.InstanceId, s_defaultName.DatabaseId, null, CancellationToken.None)
+                    .ConfigureAwait(false);
+                await pool.CreateSessionFromPoolAsync(
+                    client.Object, s_defaultName.ProjectId,
+                    s_defaultName.InstanceId, s_defaultName.DatabaseId, null, CancellationToken.None)
+                    .ConfigureAwait(false);
+
+                var releaseTask = Task.Delay(TimeSpan.FromMilliseconds(10)).ContinueWith(
+                    (x, y) => { pool.ReleaseToPool(client.Object, session1); }, null);
```

Comment:
Yes, you can do that too. You can even move them down to the bottom of the owning function if you want (you can call them before they're declared). It can go anywhere as long as the locals that are closed over have been declared (and possibly assigned, I believe) before it.